### DISCUSSION TITLE
Give feedback to the merchant and user if the tx has RBF enabled

### DISCUSSION THREAD

**Discussion by ishristov** [btcpayserver/btcpayserver#3616] [Ideas / Feature Requests]
The use case is for physical businesses that accept BTC via BTCPay and where the cashiers only have access to the Keypad POS and don't have access to the BTCPay admin panel (only the business owner has). In that case the cashiers have no idea whether the payment can turn out to be problematic (user pays with RBF payment, goes out of the store and makes a new payment with higher fee, basically robbing the store).

There are multiple businesses where the cashier can simply ask the client to wait a few minutes and to even ask them to disable the RBF next time to avoid waiting times like hotels, restaurants, self serving coffee shops, places where you pay for entry, etc.

I doubt there will be multiple people paying with BTC and RBF tx in a 10-minute timeframe, but the cashiers can open the Keypad POS multiple times on their device if needed (in different tabs), although that's a super edge case IMO.

After reading and checking all previous similar discussions my suggestion is to update the checkout body to something like that:

![feature request for rbf feedback](https://i.imgur.com/p3QqAIQ.jpeg)

I know lightning fixes this but:

- a lot of people still don't use lightning (almost no one here does) and I don't see a reason why we shouldn't improve the on-chain side as well. On-chain payments will forever be part of BTC and we shouldn't pick one over the other. We should have good solution for both types;

- some merchants prefer to use the on-chain because they want the BTC to go directly to their hardware wallets and the extra LNBank wallet is confusing for them.

**Comment by pavlenex** [comment]
To me this sounds like a reasonable use-case and request. @NicolasDorier any idea if we can do this technically?

**Comment by ishristov** [comment]
Anyway we can bump this? If anyone can put a RBF flag into the PaymentModel, I should be able to handle the UI part for the checkout myself. @NicolasDorier can we do this?

**Comment by Transisto** [comment]
Thanks for making this issue, I'd like to add all the detail of the non-confirm we check to let customers go.

Not RBF
No unconfirmed parent
Fee paid puts the transaction within one or two blocks.

I'm willing to put a 500$ worth or 2 500 000 sat bounty to get this implemented.

Until transaction on chain become prohibitive for a while, there will only be a small fraction of Bitcoin users who will seek to use lighting wallets. Also many use exchanges as a wallet, sad but the withdrawal delay on may exchange is much faster than a confirmation time. So it's usable as a method of payment.

There is a lot of demand from merchants dealing face to face to safely accept zero confirmations.

**Comment by NicolasDorier** [comment]
We shouldn't show this by default.
* Normie cashier would have no idea about what is written, and would wonder whether or not to give the good the customer bought
* Bitcoiner cashier might see this, but will you really not let the customer go if that happen? Yes if it's a car, but for a beer?
* In a non physical store, the sender might have no idea what this mean, as some wallet may not even give the ability to chose whether or nto to RBF, they will wonder why it's orange and what they did wrong.

I am open to the idea to have an option in checkout appearance for this though. But shouldn't be the default.

**Comment by ishristov** [comment]
@NicolasDorier, can we keep the current UI as is (by default), but in the checkout settings to have a new checkbox for "Show warning for RBF payment" or "Show warning for unsettled payment". If checked (and only then), the checkout page will switch the suggested UI?

This way it will give the store owner the choice to decide whether that's important or not. You are right it might make more sense in some cases and not so much in others. A new setting (and a clear checkout warning then) may solve that.

**Comment by pavlenex** [comment]
Someone proposed to me recently on a conference that we can in the future solve this on the "protocol" level, add RBF flag in BIP21 so when it's activated it can just choose not to accept the transaction.  This is harder way but imo right one, meanwhile I think was @ishristov is suggesting as an option makes total sense, we just need to get the wording right.

**Comment by Transisto** [comment]
Yes definitely should be optional.
Thinking of new cashier, I'm having concerns with btcpay makes it currently possible to show a large green checkmark when it receives a zeroconf. Zeroconf with RBF should mean very little unless there's a high trust relationship.

We currently have cashier check for the details of the transaction on a block explorer for every transaction. But we did have cases where the cashier didn't do that because the check mark looked too positive / reassuring. 

The transaction had a chain of unconfirmed parents at very low fees and took 3 days to confirm. 

On the fee side I think it's fair to show to the cashier the fee paid since the invoice suggest to the customer a specific fee to pay. 

We use mempool.space because it shows the effective fee with parents.

The flow I would see is cashier see a yellow checkmark, that reminds them to expand the transaction detail and parse the information there according to their training.

**Comment by Transisto** [comment]
Without too much work on design, How much work would it take to only show,
`*RBF` When it is one?  Not having to run a separate wallet to verify quality of zeroconf would make our life much easier.

RBF, 21sat/B, 0.24mb, Unconfirmed parent.
Or how about just letting us click a mempool space of the detected trasaction when paid but unconfirmed?

Currently it shows this empty page.
![image](https://user-images.githubusercontent.com/388716/198172857-f63379f4-07d4-4538-8c75-d251188c3eee.png)

When clicking there.
![image](https://user-images.githubusercontent.com/388716/198172905-619ca0f6-8b96-4a1a-910e-9cd08f85a4ef.png)

One page say it's paid, the next say it's not settled. Nothing more, No more address, No amounts, No transaction.
I'd be useful to be able to get more info like a TXID to be able to have the cashier or customer do some level of diagnostic without having access to the admin dashboard.

**Comment by ghost** [comment]
Any unconfirmed tx can be replaced so this warning does not make sense.

**Reply by Transisto** [reply]
A non rbf can be replaced but will not be replaced by the majority of miners currently enforcing first seen policy.

If 100% of customers are dishonest and 90% of hashrate enforce First seen policy, zeroconf have a use case for businesses with a 10% profit margin. Reality is closer to 80% of customers being honest, they have a cost for being in-person and they all benefit from not waiting an arbitrary amount of time for confirmation.
